### PR TITLE
Fichier de communication avec le stage Sutter

### PR DESCRIPTION
**Pull request by justinemajor**
Voici le fichier de communication avec le Sutter.
Here is the communication program for the Sutter stage.

### CODE REVIEW

File: hardwarelibrary/communication/sutter.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
+from struct import *
+import sys
+import serial
+from serial.tools.list_ports import comports
+#sys.path.insert(0, "/Users/justinemajor/Documents/gph.doc/git/PyHardwareLibrary/hardwarelibrary/communication")
```
Comment (dccote) [inline hardwarelibrary/communication/sutter.py:5]: Effacer.

File: hardwarelibrary/communication/sutter.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+import time
+import array
+import os
+import serial
+
+#/dev/cu.usbserial-SI8YCLBE
```
Comment (dccote) [inline hardwarelibrary/communication/sutter.py:11]: Effacer.

File: hardwarelibrary/communication/sutter.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+class SutterDevice:
+    def __init__(self):
+        """
+        SutterDevice represents a XYZ stage.  
+        """
+        
+        self.port = serial.Serial("/dev/cu.usbserial-SI8YCLBE", timeout=5, baudrate=128000)
```
Comment (dccote) [inline hardwarelibrary/communication/sutter.py:18]: There is a function to get a port based on idVendor and idProduct. They are available in SerialPort.

File: hardwarelibrary/communication/sutter.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+        #ajouter les instances nécessaires
```
Comment (dccote) [inline hardwarelibrary/communication/sutter.py:20]: Delete, use english.

File: hardwarelibrary/communication/sutter.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+    def initializeDevice(self):
+        """
+        We do a late initialization: if the device is not present at creation, it can still be
+        initialized later.
+        """
+        if self.port is not None:
+            return
+        
+        self.port = serial.Serial("/dev/cu.usbserial-SI8YCLBE", timeout=5, baudrate=128000)
```
Comment (dccote) [inline hardwarelibrary/communication/sutter.py:32]: Remove hardcoded port.

File: hardwarelibrary/communication/sutter.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+        except Exception as err:
+            print('Error when reading reply: {0}'.format(err))
+            self.shutdownDevice()
+            return None
+
+    def positionInMicrosteps(self) -> (int,int,int):
+        """ Returns the position in microsteps """
+        commandBytes = pack('<cc', b'C', b'\r')
+        self.sendCommand(commandBytes)
+        return self.readReply(size=14, format='<clllc')[1:-1]
```
Comment (dccote) [inline hardwarelibrary/communication/sutter.py:78]: This is not robust: first and last characters donèt need to be removed from the tuple.

File: hardwarelibrary/communication/sutter.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+
+    def moveInMicrostepsTo(self, position):
+        """ Move to a position in microsteps """
+        x,y,z  = position
+        commandBytes = pack('<clllc', b'M', int(x), int(y), int(z), b'\r')
+        self.sendCommand(commandBytes)
+        #self.readReply(size=1, format='<c')
```
Comment (dccote) [inline hardwarelibrary/communication/sutter.py:85]: Delete.

File: hardwarelibrary/communication/sutter.py [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+        x,y,z  = position
+        positionInMicrosteps = (x*self.microstepsPerMicrons, 
+                                y*self.microstepsPerMicrons,
+                                z*self.microstepsPerMicrons)
+        self.moveInMicrostepsTo(positionInMicrosteps)
+        #self.port.close()
+
+    def moveBy(self, delta) -> bool:
+        #Move by a delta displacement (dx, dy, dz) from current position in microns
+        #if not self.port.is_open:
```
Comment (dccote) [inline hardwarelibrary/communication/sutter.py:111]: Delete this.

File: hardwarelibrary/communication/sutter.py [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+            #self.port.open()
+        dx,dy,dz  = delta
+        position = self.position()
+        if position is not None:
+            x,y,z = position
+            self.moveTo((x+dx, y+dy, z+dz))
+        #self.port.close()
```
Comment (dccote) [inline hardwarelibrary/communication/sutter.py:118]: Delete that.

File: hardwarelibrary/communication/sutter.py [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+
+
+if __name__ == "__main__":
```
Comment (dccote) [inline hardwarelibrary/communication/sutter.py:121]: This should be elsewhere.  Where is the test file?

**Review by dccote — CHANGES_REQUESTED** [review]

File: hardwarelibrary/communication/sutter.py (hunk 3 above)
Comment (justinemajor) [reply hardwarelibrary/communication/sutter.py:13]: I can't seem to find it... Is it a function of pyserial or a module by itself?

File: hardwarelibrary/communication/sutter.py (hunk 5 above)
Comment (justinemajor) [reply hardwarelibrary/communication/sutter.py:25]: delete or replace by SerialPort (once I find it)?

File: hardwarelibrary/communication/sutter.py (hunk 10 above)
Comment (justinemajor) [reply hardwarelibrary/communication/sutter.py:107]: I can create a new file with the loop...

File: hardwarelibrary/communication/sutter.py (hunk 3 above)
Comment (dccote) [reply hardwarelibrary/communication/sutter.py:13]: Look at `sutterdevice.py` in `hardwarelibrary/motion`.

Comment (dccote) [reply hardwarelibrary/communication/sutter.py:13]:
```
from hardwarelibrary.physicaldevice import *
from hardwarelibrary.motion.linearmotiondevice import *
from hardwarelibrary.communication.communicationport import *
from hardwarelibrary.communication.commands import DataCommand

import numpy as np
import re
import time
from struct import *

class SutterDevice(PhysicalDevice, LinearMotionDevice):

    def __init__(self, bsdPath=None, portPath=None, serialNumber: str = None,
                 productId: np.uint32 = None, vendorId: np.uint32 = None):

        if bsdPath is not None:
            self.portPath = bsdPath
        elif portPath is not None:
            self.portPath = portPath
        else:
            self.portPath = None

        PhysicalDevice.__init__(self, serialNumber, vendorId, productId)
        LinearMotionDevice.__init__(self)
        self.port = None
        self.xMinLimit = 0
        self.yMinLimit = 0
        self.zMinLimit = 0
        self.xMaxLimit = 25000
        self.yMaxLimit = 25000
        self.zMaxLimit = 25000


    def __del__(self):
        try:
            self.port.close()
        except:
            # ignore if already closed
            return

    def doInitializeDevice(self): 
        try:
            if self.portPath == "debug":
                self.port = SutterDebugSerialPort()
            else:
                self.port = CommunicationPort(portPath=self.portPath)
            
            if self.port is None:
                raise PhysicalDeviceUnableToInitialize("Cannot allocate port {0}".format(self.portPath))

            self.port.open()
            self.doGetPosition()

        except Exception as error:
            if self.port is not None:
                if self.port.isOpen:
                    self.port.close()
            raise PhysicalDeviceUnableToInitialize()
        except PhysicalDeviceUnableToInitialize as error:
            raise error
        

    def doShutdownDevice(self):
        self.port.close()
        self.port = None
        return

class SutterDebugSerialPort(CommunicationPort):
    def __init__(self):
        super(SutterDebugSerialPort,self).__init__()
        self.xSteps = 0
        self.ySteps = 0
        self.zSteps = 0

        move = DataCommand(name='move', hexPattern='6d(.{8})(.{8})(.{8})')
        position = DataCommand(name='position', hexPattern='63')
        self.commands.append(move)
        self.commands.append(position)

    def processCommand(self, command, groups, inputData) -> bytearray:
        if command.name == 'move':
            self.x = unpack('<l',groups[0])[0]
            self.y = unpack('<l',groups[1])[0]
            self.z = unpack('<l',groups[2])[0]
            return b'\r'
        elif command.name == 'position':
            replyData = bytearray()
            replyData.extend(bytearray(pack("<l", self.x)))
            replyData.extend(bytearray(pack("<l", self.y)))
            replyData.extend(bytearray(pack("<l", self.z)))
            replyData.extend(b'\r')
            return replyData
```

**Pull request closed without merging**